Add endpoint to connection string

vadim2404 opened the pull request:
## Describe your changes
We should force our users to use endpoint wording vs. project.

## Issue ticket number and link
Closes neondatabase/cloud#4928

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

hlinnaka reviewed (approved):
There's an accidental change to `vendor/postgres-v15`. Other than that, +1

vadim2404 commented:
> There's an accidental change to `vendor/postgres-v15`. Other than that, +1

Yes, thanks. I didn't properly updated submodules. Thanks for highlighting

vadim2404 commented:
@SergeyMelnikov, added in these places too. Rely on @funbringer's review

funbringer reviewed (commented):
Looks good overall. I'd change a few things:

* First, extract stuff like this into a function:

```
.starts_with(b"project=") {
            // project is deprecated, but we need to have it for backward compatibility.
            bytes.strip_prefix(b"project=")
        } else {
            bytes.strip_prefix(b"endpoint=")
        }
````

E.g. `fn parse_endpoint_param(s: &str) -> Option<&str>;`.

* Add `endpoint=` variant to `test_runner/regress/test_proxy.py` tests, probably using pytest's fixtures.

vadim2404 commented:
> Looks good overall. I'd change a few things:
> 
> * First, extract stuff like this into a function:
> 
> ```
> .starts_with(b"project=") {
>             // project is deprecated, but we need to have it for backward compatibility.
>             bytes.strip_prefix(b"project=")
>         } else {
>             bytes.strip_prefix(b"endpoint=")
>         }
> ```
> 
> E.g. `fn parse_endpoint_param(s: &str) -> Option<&str>;`.
> 
> * Add `endpoint=` variant to `test_runner/regress/test_proxy.py` tests, probably using pytest's fixtures.

Addressed your comments.

bayandin commented on proxy/src/auth/credentials.rs:
```diff
@@ -61,7 +61,15 @@ impl<'a> ClientCredentials<'a> {
         // Project name might be passed via PG's command-line options.
         let project_option = params
```
Do we want to rename it to `endpoint_option`?

vadim2404 replied:
not yet, I intentionally didn't changed the whole code base

bayandin commented on proxy/src/auth/credentials.rs:
```diff
@@ -61,7 +61,15 @@ impl<'a> ClientCredentials<'a> {
         let project_option = params
             .options_raw()
-            .and_then(|mut options| options.find_map(|opt| opt.strip_prefix("project=")))
+            .and_then(|mut options| {
+                options.find_map(|opt| {
+                    if opt.starts_with("project=") {
+                        opt.strip_prefix("project=")
+                    } else {
+                        opt.strip_prefix("endpoint=")
+                    }
+                })
+            })
```
What will happen if someone passes both of these parameters in the options?

bayandin replied:
> Do you think we need to abrupt the connection if two options exist?

I think it should use `endpoint` parameter if it exists, if not, then check `project` as a fallback.
Now it looks like some undefined behaviour, I'd like to have it consistent (an option with abrupting connection suits too)

funbringer commented on test_runner/fixtures/neon_fixtures.py:
```diff
@@ -2237,6 +2237,19 @@ def static_proxy(
+        "endpoint",
+    ],
+)
+def proxy_endpoint_option_name(request: FixtureRequest) -> str:
+    assert hasattr(request, "param")
+    assert isinstance(request.param, str)
+    return request.param
```
To me putting this small piece in the fixtures library doesn't make much sense. It's very ad-hoc and hard to discover.

 I'd solve this using `pytest.mark.parametrize`:

```python
@pytest.mark.parametrize("name", ["endpoint", "project"])
def test_fizz_buzz(name: str):
    pass
```

vadim2404 replied:
It was my initial implementation, btw. But I followed your idea from initial comment

funbringer replied:
Sorry if I gave a wrong impression; this is what I had in mind at that time.

funbringer commented on proxy/src/auth/credentials.rs:
```diff
@@ -61,7 +61,21 @@ impl<'a> ClientCredentials<'a> {
+                // Therefore we require that only one of them is present.
+                options
+                    .filter_map(|opt| {
+                        if opt.starts_with("project=") {
+                            opt.strip_prefix("project=")
+                        } else {
+                            opt.strip_prefix("endpoint=")
+                        }
```
it's `parse_endpoint_param`

funbringer commented on proxy/src/compute.rs:
```diff
@@ -280,6 +280,7 @@ fn filtered_options(params: &StartupMessageParams) -> Option<String> {
     let options: String = params
         .options_raw()?
         .filter(|opt| !opt.starts_with("project="))
+        .filter(|opt| !opt.starts_with("endpoint="))
```
`filter(|opt| parse_endpoint_param(opt).is_none())`

funbringer commented on proxy/src/auth/password_hack.rs:
```diff
@@ -6,18 +6,27 @@
+}
+
+fn parse_endpoint_param(bytes: &[u8]) -> Option<&[u8]> {
+    if bytes.starts_with(b"project=") {
+        // project is deprecated, but we need to have it for backward compatibility.
+        bytes.strip_prefix(b"project=")
+    } else {
+        bytes.strip_prefix(b"endpoint=")
     }
```
```rust
s.strip_prefix(b"endpoint=").or_else(|| s.strip_prefix(b"project="))
```
